File: TL.py
__author__ = 'agrigoryev'
import os
import struct
import json
import io
from numbers import Number

class TlConstructor:
    def __init__(self, json_dict):
        self.id = int(json_dict['id'])
        self.type = json_dict['type']
        self.predicate = json_dict['predicate']
        self.params = []
        # case of vector
        for param in json_dict['params']:
            if param['type'] == "Vector<long>":
                param['type'] = "Vector t"
                param['subtype'] = "long"
            elif param['type'] == "vector<%Message>":
                param['type'] = "vector"
                param['subtype'] = "message"
            elif param['type'] == "vector<future_salt>":
                param['type'] = "vector"
                param['subtype'] = "future_salt"
            else:
                param['subtype'] = None
            self.params.append(param)

class TlMethod:
    def __init__(self, json_dict):
        self.id = int(json_dict['id'])
        self.type = json_dict['type']
        self.method = json_dict['method']
        self.params = json_dict['params']


class TLObject(dict):
    def __init__(self, tl_elem):
        self.name = tl_elem.predicate

class TL:
    def __init__(self, filename):
        with open(filename, 'r') as f:
           TL_dict = json.load(f)

        # Read constructors

        self.constructors = TL_dict['constructors']
        self.constructor_id = {}
        self.constructor_type = {}
        for elem in self.constructors:
            z = TlConstructor(elem)
            self.constructor_id[z.id] = z
            self.constructor_type[z.predicate] = z

        self.methods = TL_dict['methods']
        self.method_id = {}
        self.method_name = {}
        for elem in self.methods:
            z = TlMethod(elem)
            self.method_id[z.id] = z
            self.method_name[z.method] = z


## Loading TL_schema (should be placed in the same directory as mtproto.py
tl = TL(os.path.join(os.path.dirname(__file__), "TL_schema.JSON"))


def serialize_obj(type_, **kwargs):
    bytes_io = io.BytesIO()
    try:
        tl_constructor = tl.constructor_type[type_]
    except KeyError:
        raise Exception("Could not extract type: %s" % type_)
    bytes_io.write(struct.pack('<i', tl_constructor.id))
    for arg in tl_constructor.params:
        serialize_param(bytes_io, type_=arg['type'],  value=kwargs[arg['name']])
    return bytes_io.getvalue()


def serialize_method(type_, **kwargs):
    bytes_io = io.BytesIO()
    try:
        tl_method = tl.method_name[type_]
    except KeyError:
        raise Exception("Could not extract type: %s" % type_)
    bytes_io.write(struct.pack('<i', tl_method.id))
    for arg in tl_method.params:
        serialize_param(bytes_io, type_=arg['type'], value=kwargs[arg['name']])
    return bytes_io.getvalue()


def serialize_param(bytes_io, type_, value):
    if type_ == "int":
        assert isinstance(value, Number)
        assert value.bit_length() <= 32
        bytes_io.write(struct.pack('<i', value))
    elif type_ == "long":
        assert isinstance(value, Number)
        bytes_io.write(struct.pack('<q', value))
    elif type_ in ["int128", "int256"]:
        assert isinstance(value, bytes)
        bytes_io.write(value)
    elif type_ == 'string' or 'bytes':
        l = len(value)
        if l < 254: # short string format
            bytes_io.write(struct.pack('<b', l))  # 1 byte of string
            bytes_io.write(value)   # string
            bytes_io.write(b'\x00'*((-l-1) % 4))  # padding bytes
        else:
            bytes_io.write(b'\xfe')  # byte 254
            bytes_io.write(struct.pack('<i', l)[:3])  # 3 bytes of string
            bytes_io.write(value) # string
            bytes_io.write(b'\x00'*(-l % 4))  # padding bytes

def deserialize(bytes_io, type_=None, subtype=None):
    """
    :type bytes_io: io.BytesIO object
    """
    assert isinstance(bytes_io, io.BytesIO)

    # Built-in bare types
    if   type_ == 'int':    x = struct.unpack('<i', bytes_io.read(4))[0]
    elif type_ == '#':      x = struct.unpack('<I', bytes_io.read(4))[0]
    elif type_ == 'long':   x = struct.unpack('<q', bytes_io.read(8))[0]
    elif type_ == 'double': x = struct.unpack('<d', bytes_io.read(8))[0]
    elif type_ == 'int128': x = bytes_io.read(16)
    elif type_ == 'int256': x = bytes_io.read(32)
    elif type_ == 'string' or type_ == 'bytes':
        l = struct.unpack('<B', bytes_io.read(1))[0]
        assert l <= 254  # In general, 0xFF byte is not allowed here
        if l == 254:
            # We have a long string
            long_len = struct.unpack('<I', bytes_io.read(3)+b'\x00')[0]
            x = bytes_io.read(long_len)
            bytes_io.read(-long_len % 4)  # skip padding bytes
        else:
            # We have a short string
            x = bytes_io.read(l)
            bytes_io.read(-(l+1) % 4)  # skip padding bytes
        assert isinstance(x, bytes)
    elif type_ == 'vector':
        assert subtype is not None
        count = struct.unpack('<l', bytes_io.read(4))[0]
        x = [deserialize(bytes_io, type_=subtype) for i in range(count)]
    else:
        # known types
        try:
        # Bare types
            tl_elem = tl.constructor_type[type_]
        except KeyError:
            # Boxed types
            i = struct.unpack('<i', bytes_io.read(4))[0]  # read type ID
            try:
                tl_elem = tl.constructor_id[i]
            except KeyError:
                # Unknown type
                raise Exception("Could not extract type: %s" % type_)

        base_boxed_types = ["Vector t", "Int", "Long", "Double", "String", "Int128", "Int256"]
        if tl_elem.type in base_boxed_types:
            x = deserialize(bytes_io, type_=tl_elem.predicate, subtype=subtype)
        else:  # other types
            x = TLObject(tl_elem)
            for arg in tl_elem.params:
                x[arg['name']] = deserialize(bytes_io, type_=arg['type'], subtype=arg['subtype'])
    return x

File: TL.php
<?php

set_include_path(get_include_path().PATH_SEPARATOR.dirname(__FILE__).DIRECTORY_SEPARATOR.'libpy2php');
require_once 'libpy2php.php';
$__author__ = 'agrigoryev';
require_once 'os.php';
class TlConstructor
{
    public function __construct($json_dict)
    {
        $this->id = (int) $json_dict['id'];
        $this->type = $json_dict['type'];
        $this->predicate = $json_dict['predicate'];
        $this->params = [];
        foreach ($json_dict['params'] as $param) {
            if (($param['type'] == 'Vector<long>')) {
                $param['type'] = 'Vector t';
                $param['subtype'] = 'long';
            } elseif (($param['type'] == 'vector<%Message>')) {
                $param['type'] = 'vector';
                $param['subtype'] = 'message';
            } elseif (($param['type'] == 'vector<future_salt>')) {
                $param['type'] = 'vector';
                $param['subtype'] = 'future_salt';
            } else {
                $param['subtype'] = null;
            }
            $this->params[] = $param;
        }
    }
}
class TlMethod
{
    public function __construct($json_dict)
    {
        $this->id = (int) $json_dict['id'];
        $this->type = $json_dict['type'];
        $this->method = $json_dict['method'];
        $this->params = $json_dict['params'];
    }
}
class TLObject extends ArrayObject
{
    public function __construct($tl_elem)
    {
        parent::__construct();
        $this->name = $tl_elem->predicate;
    }
}
class TL
{
    public function __construct($filename)
    {
        $TL_dict = json_decode(file_get_contents($filename), true);
        $this->constructors = $TL_dict['constructors'];
        $this->constructor_id = [];
        $this->constructor_type = [];
        foreach ($this->constructors as $elem) {
            $z = new TlConstructor($elem);
            $this->constructor_id[$z->id] = $z;
            $this->constructor_type[$z->predicate] = $z;
        }
        $this->methods = $TL_dict['methods'];
        $this->method_id = [];
        $this->method_name = [];
        foreach ($this->methods as $elem) {
            $z = new TlMethod($elem);
            $this->method_id[$z->id] = $z;
            $this->method_name[$z->method] = $z;
        }
        $this->struct = new \danog\PHP\Struct();
    }

    public function serialize_obj($type_, $kwargs)
    {
        $bytes_io = fopen('php://memory', 'rw+b');
        if (isset($this->constructor_type[$type_])) {
            $tl_constructor = $this->constructor_type[$type_];
        } else {
            throw new Exception(sprintf('Could not extract type: %s', $type_));
        }
        fwrite($bytes_io, $this->struct->pack('<i', $tl_constructor->id));
        foreach ($tl_constructor->params as $arg) {
            $this->serialize_param($bytes_io, $arg['type'], $kwargs[$arg['name']]);
        }

        return fread_all($bytes_io);
    }

    public function serialize_method($type_, $kwargs)
    {
        $bytes_io = fopen('php://memory', 'rw+b');
        if (isset($this->method_name[$type_])) {
            $tl_method = $this->method_name[$type_];
        } else {
            throw new Exception(sprintf('Could not extract type: %s', $type_));
        }
        fwrite($bytes_io, $this->struct->pack('<i', $tl_method->id));
        foreach ($tl_method->params as $arg) {
            $this->serialize_param($bytes_io, $arg['type'], $kwargs[$arg['name']]);
        }

        return fread_all($bytes_io);
    }

    public function serialize_param($bytes_io, $type_, $value)
    {
        switch ($type_) {
            case 'int':
                assert(is_int($value));
                assert(strlen(decbin($value)) <= 32);
                fwrite($bytes_io, $this->struct->pack('<i', $value));
                break;
            case 'long':
                assert(is_long($value));
                fwrite($bytes_io, $this->struct->pack('<q', $value));
                break;
            case 'int128':
            case 'int256':
                assert(is_string($value));
                fwrite($bytes_io, $value);
                break;
            case 'string':
            case 'bytes':
                $l = len($value);
                if (($l < 254)) {
                    fwrite($bytes_io, $this->struct->pack('<b', $l));
                    fwrite($bytes_io, $value);
                    fwrite($bytes_io, pack('@'.posmod((-$l - 1), 4)));
                } else {
                    fwrite($bytes_io, string2bin('\xfe'));
                    fwrite($bytes_io, substr($this->struct->pack('<i', $l), 0, 3));
                    fwrite($bytes_io, $value);
                    fwrite($bytes_io, pack('@'.posmod(-$l, 4)));
                }
                break;
            default:
                break;
        }
    }

    /**
     * :type bytes_io: io.BytesIO object.
     */
    public function deserialize($bytes_io, $type_ = null, $subtype = null)
    {
        assert(get_resource_type($bytes_io) == 'file' || get_resource_type($bytes_io) == 'stream');
        switch ($type_) {
            case 'int':
                $x = $this->struct->unpack('<i', fread($bytes_io, 4)) [0];
                break;
            case '#':
                $x = $this->struct->unpack('<I', fread($bytes_io, 4)) [0];
                break;
            case 'long':
                $x = $this->struct->unpack('<q', fread($bytes_io, 8)) [0];
                break;
            case 'double':
                $x = $this->struct->unpack('<d', fread($bytes_io, 8)) [0];
                break;
            case 'int128':
                $x = fread($bytes_io, 16);
                break;
            case 'int256':
                $x = fread($bytes_io, 32);
                break;
            case 'string':
            case 'bytes':
                $l = $this->struct->unpack('<B', fread($bytes_io, 1)) [0];
                assert($l <= 254);
                if (($l == 254)) {
                    $long_len = $this->struct->unpack('<I', fread($bytes_io, 3).string2bin('\x00')) [0];
                    $x = fread($bytes_io, $long_len);
                    fread($bytes_io, posmod(-$long_len, 4));
                } else {
                    $x = fread($bytes_io, $l);
                    fread($bytes_io, posmod(-($l + 1), 4));
                }
                assert(is_string($x));
                break;
            case 'vector':
                assert($subtype != null);
                $count = $this->struct->unpack('<l', fread($bytes_io, 4)) [0];
                $x = [];
                foreach (pyjslib_range($count) as $i) {
                    $x[] = $this->deserialize($bytes_io, $subtype);
                }
                break;
            default:
                if (isset($this->constructor_type[$type_])) {
                    $tl_elem = $this->constructor_type[$type_];
                } else {
                    $Idata = fread($bytes_io, 4);
                    $i = $this->struct->unpack('<i', $Idata) [0];
                    if (isset($this->constructor_id[$i])) {
                        $tl_elem = $this->constructor_id[$i];
                    } else {
                        throw new Exception(sprintf('Could not extract type: %s', $type_));
                    }
                }

                $base_boxed_types = ['Vector t', 'Int', 'Long', 'Double', 'String', 'Int128', 'Int256'];
                if (in_array($tl_elem->type, $base_boxed_types)) {
                    $x = $this->deserialize($bytes_io, $tl_elem->predicate, $subtype);
                } else {
                    $x = new TLObject($tl_elem);
                    foreach ($tl_elem->params as $arg) {
                        $x[$arg['name']] = $this->deserialize($bytes_io, $arg['type'], $arg['subtype']);
                    }
                }
                break;
        }

        return $x;
    }
}
